internal/dcache/rpc/server/handler.go

```diff
@@ -47,43 +57,894 @@ var _ service.ChunkService = &ChunkServiceHandler{}
+	for _, info := range h.rvIDMap {
+		myRvList = append(myRvList, info.rvName)
+	}
+
+	return &models.HelloResponse{
+		ReceiverNodeID: myNodeID,
+		Time:           time.Now().UnixMicro(),
+		RV:             myRvList,
+		MV:             req.MV, // TODO:: discuss: how to fetch the MV list receiver shares with the sender
+	}, nil
 }
 
 func (h *ChunkServiceHandler) GetChunk(ctx context.Context, req *models.GetChunkRequest) (*models.GetChunkResponse, error) {
-	return nil, nil
+	if req == nil || req.Address == nil {
+		log.Err("ChunkServiceHandler::GetChunk: Received nil GetChunk request")
+		common.Assert(false, "received nil GetChunk request")
+		return nil, rpc.NewResponseError(rpc.InvalidRequest, "received nil GetChunk request")
+	}
+
+	log.Debug("ChunkServiceHandler::GetChunk: Received GetChunk request: %+v", *req)
+
+	// check if the chunk address is valid
+	err := h.checkValidChunkAddress(req.Address)
+	if err != nil {
+		log.Err("ChunkServiceHandler::GetChunk: Invalid chunk address [%s]", err.Error())
+		return nil, err
+	}
+
+	rvInfo := h.rvIDMap[req.Address.RvID]
+	mvInfo, _ := rvInfo.getMVInfo(req.Address.MvName)
+
+	// check if the sync has started. If yes, block the chunk operations till the sync is completed
+	mvInfo.blockChunkOps()
+
+	// increment the chunk operation count for this MV
+	mvInfo.incrementChunkOps()
+
+	// decrement the chunk operation count for this MV when the function returns
+	defer mvInfo.decrementChunkOps()
+
+	startTime := time.Now()
+
+	// check if the chunk file is being updated in parallel by some other thread
+	chunkAddress := getChunkAddress(req.Address.FileID, req.Address.RvID, req.Address.MvName, req.Address.OffsetInMB)
+	flock := h.locks.Get(chunkAddress)
+	flock.Lock()
+	defer flock.Unlock()
+
+	cacheDir := rvInfo.cacheDir
+	chunkPath, hashPath := getChunkAndHashPath(cacheDir, req.Address.MvName, req.Address.FileID, req.Address.OffsetInMB)
+	log.Debug("ChunkServiceHandler::GetChunk: chunk path %s, hash path %s", chunkPath, hashPath)
+
+	fh, err := os.Open(chunkPath)
+	if err != nil {
+		log.Err("ChunkServiceHandler::GetChunk: Failed to open chunk file %s [%v]", chunkPath, err.Error())
+		return nil, rpc.NewResponseError(rpc.InternalServerError, fmt.Sprintf("failed to open chunk file %s [%v]", chunkPath, err.Error()))
+	}
+	defer fh.Close()
+
+	// TODO:: integration: call cluster manager to get chunk size
+	chunkSize := getChunkSize()
+
+	if req.Length == -1 {
+		req.Length = chunkSize - req.Offset
+	}
+
+	data := make([]byte, req.Length)
+	_, err = fh.ReadAt(data, req.Offset)
+	// TODO: should we handle EOF error
+	if err != nil {
+		log.Err("ChunkServiceHandler::GetChunk: Failed to read chunk file %s [%v]", chunkPath, err.Error())
+		return nil, rpc.NewResponseError(rpc.InternalServerError, fmt.Sprintf("failed to read chunk file %s [%v]", chunkPath, err.Error()))
+	}
+
+	// get hash if requested for entire chunk
+	hash := ""
+	if req.Offset == 0 && req.Length == chunkSize {
```

Comment:
req.Length == chunkSize is not the right check for last chunks which may not be full size.
Better check would be to see if we read the full chunk from offset=0 till the end. For this we can use the file size calculated above using a stat() call.

Reply:
req.Length is updated to chunkSize above using,
```
chunkSize := fInfo.Size() // got from stat

if req.Length == -1 {
	common.Assert(chunkSize > req.Offset, fmt.Sprintf("chunkSize %d is less than req.Offset %d", chunkSize, req.Offset))
	req.Length = chunkSize - req.Offset
}
```

So, this logic will work for last chunk also